File: scripts/generate-embeddings.js
// Embedding Generation Script
// Pre-computes embeddings for all error codes using Transformers.js
// Run with: node scripts/generate-embeddings.js

const fs = require('fs');
const path = require('path');

// Check if transformers is available
let pipeline;
try {
    pipeline = require('@xenova/transformers').pipeline;
} catch (e) {
    console.log('⚠️  @xenova/transformers not installed globally');
    console.log('   Trying to use local installation...');
    try {
        require('./node_modules/@xenova/transformers');
        pipeline = require('./node_modules/@xenova/transformers').pipeline;
    } catch (e2) {
        console.error('❌ Failed to load transformers. Please install:');
        console.error('   npm install @xenova/transformers');
        process.exit(1);
    }
}

async function generateEmbeddings() {
    console.log('🚀 Starting embedding generation...');
    
    // Load error codes data
    const dataPath = path.join(__dirname, '..', 'data', 'error_codes.json');
    const data = JSON.parse(fs.readFileSync(dataPath, 'utf8'));
    
    const errorCodes = data.errorCodes;
    console.log(`📊 Loaded ${errorCodes.length} error codes`);
    
    // Load transformer model
    console.log('📥 Loading multilingual model: paraphrase-multilingual-MiniLM-L12-v2');
    const embedder = await pipeline('feature-extraction', 'Xenova/paraphrase-multilingual-MiniLM-L12-v2');
    console.log('✅ Model loaded');
    
    const embeddings = [];
    const batchSize = 32;
    
    for (let i = 0; i < errorCodes.length; i += batchSize) {
        const batch = errorCodes.slice(i, i + batchSize);
        const texts = batch.map(item => item.description);
        
        // Generate embeddings for batch
        const results = await embedder(texts, { pooling: 'mean', normalize: true });
        
        batch.forEach((item, idx) => {
            const embedding = Array.from(results[idx]);
            embeddings.push({
                code: item.code,
                description: item.description,
                category: item.category || '',
                embedding: embedding
            });
        });
        
        const progress = Math.min(i + batchSize, errorCodes.length);
        console.log(`📈 Progress: ${progress}/${errorCodes.length} (${Math.round(progress/errorCodes.length*100)}%)`);
    }
    
    // Save embeddings
    const outputPath = path.join(__dirname, '..', 'embeddings', 'error_codes_embedded.json');
    fs.writeFileSync(outputPath, JSON.stringify(embeddings, null, 0));
    
    console.log(`✅ Saved ${embeddings.length} embeddings to ${outputPath}`);
    console.log(`   Embedding dimension: ${embeddings[0]?.embedding?.length || 'unknown'}`);
    
    // Also create a metadata file
    const metadata = {
        model: 'Xenova/paraphrase-multilingual-MiniLM-L12-v2',
        embedding_dim: embeddings[0]?.embedding?.length || 384,
        count: embeddings.length,
        created_at: new Date().toISOString()
    };
    fs.writeFileSync(
        path.join(__dirname, '..', 'embeddings', 'metadata.json'),
        JSON.stringify(metadata, null, 2)
    );
    
    console.log('🎉 Done!');
}

generateEmbeddings().catch(console.error);

File: scripts/generate_embeddings.py
#!/usr/bin/env python3
"""
Embedding Generation Script
Pre-computes embeddings for all error codes using sentence-transformers
Run with: python3 scripts/generate_embeddings.py
"""

import json
import os
import sys

# Check for sentence-transformers
try:
    from sentence_transformers import SentenceTransformer
except ImportError:
    print("❌ sentence-transformers not installed")
    print("   Run: pip install sentence-transformers")
    sys.exit(1)

def generate_embeddings():
    print("🚀 Starting embedding generation...")
    
    # Load error codes data
    data_path = os.path.join(os.path.dirname(__file__), '..', 'data', 'error_codes.json')
    with open(data_path, 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    error_codes = data['errorCodes']
    print(f"📊 Loaded {len(error_codes)} error codes")
    
    # Load multilingual model
    model_name = 'sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2'
    print(f"📥 Loading multilingual model: {model_name}")
    model = SentenceTransformer(model_name)
    print("✅ Model loaded")
    
    # Generate embeddings
    texts = [item['description'] for item in error_codes]
    print(f"🔄 Generating embeddings for {len(texts)} descriptions...")
    embeddings_list = model.encode(texts, show_progress_bar=True, batch_size=32)
    
    # Create output data
    embeddings = []
    for i, item in enumerate(error_codes):
        embeddings.append({
            'code': item['code'],
            'description': item['description'],
            'category': item.get('category', ''),
            'embedding': embeddings_list[i].tolist()
        })
    
    # Save embeddings
    output_path = os.path.join(os.path.dirname(__file__), '..', 'embeddings', 'error_codes_embedded.json')
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(embeddings, f)
    
    print(f"✅ Saved {len(embeddings)} embeddings to {output_path}")
    print(f"   Embedding dimension: {len(embeddings[0]['embedding']) if embeddings else 'unknown'}")
    
    # Also create a metadata file
    metadata = {
        'model': model_name,
        'embedding_dim': len(embeddings[0]['embedding']) if embeddings else 384,
        'count': len(embeddings),
        'created_at': __import__('datetime').datetime.now().isoformat()
    }
    
    metadata_path = os.path.join(os.path.dirname(__file__), '..', 'embeddings', 'metadata.json')
    with open(metadata_path, 'w', encoding='utf-8') as f:
        json.dump(metadata, f, indent=2)
    
    print("🎉 Done!")

if __name__ == '__main__':
    generate_embeddings()